### PR TITLE
Proof development primitives

### PR DESCRIPTION
**Pull request by robdockins**
This PR reorganizes the execution internals of SAWScript and implements some additional experimental primitives that will hopefully yield quality-of-life improvements for proof development.

The two main new capabilities are a `callcc` primitive, and a `subshell` primitive. The `subshell` primitive, when executed, will drop the user directly into a REPL session with the same context as the point where `subshell` was called. This should allow the user to experiment, view internal state, etc. interactively. `callcc`, as typical, reifies the rest of the computation as a function that can be called.  This is especially useful in combination with `subshell`, as the user can experiment with state updates and see what effect that has on the remainder of the proof without having to rerun the proof from the beginning.

There is additionally a new `checkpoint` primitive that captures (some) of the mutable state of the system and allows it to be restored at a later time. This facility is _highly_ experimental, and it is relatively easy to work yourself into broken states using it. Probably, we need a more comprehensive plan to deal with the general idea of checkpoint/resume in the entire SAW ecosystem to make this work more reliably.

### CODE REVIEW

**Comment by robdockins** [comment]
Here is a quick example of what this makes possible.  Consider the following SAW script, and the following interaction.

```
enable_experimental;

let interactive =
  callcc (\ (done : Int -> TopLevel ()) ->
      rec retry (u: ()) : TopLevel Int =
        do{ reset <- checkpoint;
	    print "<<Interactive subshell>>";
            subshell ();
	    reset ();
	    retry (); }
      in retry ());

x <- interactive;

print (str_concat "x = " (show x));
print (str_concat "y = " (show y));
```


```
$ saw interactive.saw


[17:28:17.889] Loading file "/Users/rdockins/code/saw-script/interactive.saw"
[17:28:17.890] <<Interactive subshell>>
sawscript> let y = 10
sawscript> done 42
[17:28:28.472] x = 42
[17:28:28.472] y = 10
sawscript> let y = 12
sawscript> done 100
[17:28:37.856] x = 100
[17:28:37.856] y = 12
sawscript> ^D
[17:28:39.064] Restoring state from checkpoint
[17:28:39.064] <<Interactive subshell>>
sawscript> done 100
[17:28:41.496] x = 100

/Users/rdockins/code/saw-script/interactive.saw:16:27-16:33: Unbound variable: "y" (/Users/rdockins/code/saw-script/interactive.saw:16:32-16:33)
Note that some built-in commands are available only after running
either `enable_deprecated` or `enable_experimental`.

sawscript> exit 0
```

**Comment by chameco** [comment]
This is really neat! Some thoughts before looking at the code:

I've been thinking a lot recently about to-disk serialization for `Term`s, and maybe even of the SAW context at large. It seems like at least some of the work to make `checkpoint` or continuations less broken would also make this serialization more doable - things like making sure important state lives somewhere accessible instead of hidden inside a closure somewhere, and solving problems related to moving `Term`s between `SharedContext`s.

We should also think about the Python story. I think a debugging tool like `subshell` will be very useful, but it will further widen the (already substantial in my opinion) user experience gap between SAWScript and the Python interface.

File: src/SAWScript/Builtins.hs [hunk 1]
```diff
@@ -203,11 +203,13 @@ readSBV path unintlst =
 -- support for latches in the 'AIGNetwork' SAWScript type.
 dsecPrint :: TypedTerm -> TypedTerm -> TopLevel ()
 dsecPrint t1 t2 = do
-  withSystemTempFile ".aig" $ \path1 _handle1 -> do
-  withSystemTempFile ".aig" $ \path2 _handle2 -> do
-  Prover.writeSAIGInferLatches path1 t1
-  Prover.writeSAIGInferLatches path2 t2
-  io $ callCommand (abcDsec path1 path2)
+  write_t1 <- Prover.writeSAIGInferLatches t1
```
Comment (chameco) [inline src/SAWScript/Builtins.hs:206]: I think these AIG-related changes might be unrelated work that was caught in the first commit.

File: intTests/test_external_abc/test.saw [hunk 2]
```diff
@@ -44,8 +44,8 @@ let order_res = {{ ([[0x02,0x03],[0x04,0x05]], 0x81050fff) }};
 
 // Check that Verilog counterexamples are in the right order
 sr1 <- sat w4_abc_verilog order_term;
-caseSatResult sr1 undefined (\t -> prove_print yices {{ t == order_res }});
+caseSatResult sr1 (fail "ABC verilog sat fail") (\t -> prove_print yices {{ t == order_res }});
```
Comment (chameco) [inline intTests/test_external_abc/test.saw:47]: Maybe an unrelated change?

File: src/SAWScript/Interpreter.hs [hunk 3]
```diff
@@ -797,6 +783,36 @@ primitives = Map.fromList
...
+    , "NOTE that this facility is highly experimental and may not"
+    , "be entirely reliable.  It is intended only for proof development"
+    , "where it can speed up the process of experimenting with"
+    , "mid-proof changes. Finalized proofs should not use this facility."
+    ]
+
+  , prim "subshell"            "() -> TopLevel ()"
+    (\_ _ -> toplevelSubshell)
+    Experimental
+    []
```
Comment (chameco) [inline src/SAWScript/Interpreter.hs:814]: We should make sure to include documentation for `subshell` before merging.

File: src/SAWScript/Prover/Exporter.hs [hunk 4]
```diff
@@ -203,13 +203,16 @@ writeSAIG file tt numLatches = do
 
 -- | Given a term a type '(i, s) -> (o, s)', call @writeSAIG@ on term
 -- with latch bits set to '|s|', the width of 's'.
-writeSAIGInferLatches :: FilePath -> TypedTerm -> TopLevel ()
-writeSAIGInferLatches file tt = do
+writeSAIGInferLatches :: TypedTerm -> TopLevel (FilePath -> IO ())
```
Comment (chameco) [inline src/SAWScript/Prover/Exporter.hs:206]: Maybe an unrelated change?

**Comment by robdockins** [comment]
> This is really neat! Some thoughts before looking at the code:
> 
> I've been thinking a lot recently about to-disk serialization for `Term`s, and maybe even of the SAW context at large. It seems like at least some of the work to make `checkpoint` or continuations less broken would also make this serialization more doable - things like making sure important state lives somewhere accessible instead of hidden inside a closure somewhere, and solving problems related to moving `Term`s between `SharedContext`s.

Agreed, these are all important and useful problems to be addressing, and should help with this kind of brokeness.

> We should also think about the Python story. I think a debugging tool like `subshell` will be very useful, but it will further widen the (already substantial in my opinion) user experience gap between SAWScript and the Python interface.

Also agreed.  In some ways, I feel like the Python story is already ahead here, but the user-facing interaction needs some work.

File: intTests/test_external_abc/test.saw (hunk 2 above)
Comment (robdockins) [reply intTests/test_external_abc/test.saw:47]: This turned out to be necessary to fix the test.  I think the change in the `TopLevel` monad had some subtle order-of-evaluation changes that made `undefined` evaluate later.

File: src/SAWScript/Builtins.hs (hunk 1 above)
Comment (robdockins) [reply src/SAWScript/Builtins.hs:206]: This is actually related, although the reason is pretty subtle.  `TopLevel` no longer has a `MonadMask` instance, as continuation-based monads don't provide the necessary guarantees. However `withSystemTempFile` demands `MonadMask` in order to manage the temporary file lifetime. So, one solution is this, where we extract an IO action from the `TopLevel` and execute it later in an IO context.

This seems to be the only place that `MonadMask` instance was being used.

**Comment by robdockins** [comment]
Is this what it looks like when we hit the CI memory limits?

```
[20:55:12.365] Branching on 1 override variants of OPENSSL_malloc ...
[20:55:12.365] Applied override! OPENSSL_malloc
[20:55:12.366] Symbolic simulation completed with side conditions.
[20:55:12.399] Checking proof obligations ECDSA_do_sign ...
Killed
137
make: *** [awslc] Error 137
Makefile:25: recipe for target 'awslc' failed
```

**Comment by RyanGlScott** [comment]
> Is this what it looks like when we hit the CI memory limits?

Most likely. I've seen the same error code (137) in the CI for #1563 as well.

**Comment by robdockins** [comment]
According to my local readings, the memory usage of this branch is only very slightly higher than for `master`, but I guess we are right on that knife edge.

**Comment by RyanGlScott** [comment]
FYI, the CI on this is not running due to a merge conflict (most likely with #1695).

**Review by chameco — APPROVED** [review]
LGTM, although it looks like there's some weirdness happening with the macOS CI. I can grab and unzip the supposedly problematic zip file, so maybe just an Actions hiccup?

**Comment by robdockins** [comment]
Logs indicate some kind of failure downloading `what4-solvers`...

**Comment by robdockins** [comment]
Whelp... looks like we're failing on memory limits again.

**Comment by robdockins** [comment]
After a rebase, this is back under memory limits... I think I'm going to go ahead and merge it, even though the CI seems a bit erratic.
